### DISCUSSION TITLE
BeiDou satellites?

### DISCUSSION THREAD

**Discussion by szilvasyz** [semuconsulting/pynmeagps#41] [General]
Hello,

I am helping in an Arduino project for one of my friends. We use GPS/GNSS modules in this project and I found this amazing software for PC for diagnosing the output of modules.

Some of our cheap chinese GNSS modules seem sending information on BeiDou satellites with BDGSV sentences instead of the common (or standard?) GBGSV sentences. Simply said they use the talker ID "BD" of BeiDou positioning system. The uncommon NMEA sentences even not appear in the console window, they only can be observed using a regular serial port monitor.

Due to the weird messages, there could be appear a very strange state in the number of satellites: more satellites are active than followed. Is there any possibility to make PyGPSClient know that talker ID "BD" is a synonym of "GB" to handle them identically?

Thank you for answering in advance.

**Comment by semuadmin** [comment]
Hi @szilvasyz 

So, as you're aware, the offiical NMEA 4.0 talker ID for the Beidou constellation is "GB" - "BD" is a legacy code which is effectively obsolete, but it would be perfectly possible to add support for it and I'm happy to do so. Assuming the messages are otherwise standard, this would require a simple change to the `pynmeagps` library which PyGPSClient uses to parse NMEA messages. I'll copy this request over to the `pynmeagps` GitHub repo for tracking purposes.

It would greatly assist me if you could send me a binary dump of the output of one of your cheap Chinese GNSS module containing these "BD" sentences - none of the receivers I have output it. If you don't want to post it publicly in this Discussion, you can send it to me directly at semuadmin@semuconsulting.com.

**Reply by szilvasyz** [reply]
![kép](https://github.com/semuconsulting/PyGPSClient/assets/31009531/55a9ffdc-ab46-4561-afde-65d2715d07da)
[beidou_nmea.log](https://github.com/semuconsulting/PyGPSClient/files/12806948/beidou_nmea.log)
Hi @Semuadmin,

Thank you very much for your positive words on making efforts to implement the support of this obsolete format.

I am attaching a sample log with the BDGSV sentences. On the screenshot you can see that lines seem regular NMEA sentences, line endings are CR+LF, and I have checked some NMEA checksum, those were correct. The sentences containing my geolocation have been discarded from the file. 

I have double checked the file directly logged from the terminal (TeraTerm) with a hex file editor (HxD), the line endings are CR+LF also before and after the editing and no invisible characters in the file. I hope, uploading will keep the original format.

Please tell me if I can help the realizing with anything else from my side.

Regards,

**Reply by semuadmin** [reply]
Hi @szilvasyz 

Thanks for the test file.

So I've made the change to `pynmeagps` and it can now successfully parse BDGSV sentences, e.g. using the `nmeafile.py` example:

```
python3 nmeafile.py

Enter fully qualified name of file containing raw NMEA data: /Users/semuadmin/Downloads/beidou_nmea.log

...

<NMEA(GPGSV, numMsg=3, msgNum=3, numSV=10, svid_01=23, elv_01=14.0, az_01=272, cno_01=18, svid_02=24, elv_02=69.0, az_02=301, cno_02=31, signalID=0)>
<NMEA(BDGSV, numMsg=2, msgNum=1, numSV=8, svid_01=14, elv_01=5.0, az_01=191, cno_01=, svid_02=24, elv_02=56.0, az_02=75, cno_02=25, svid_03=25, elv_03=61.0, az_03=287, cno_03=25, svid_04=26, elv_04=9.0, az_04=87, cno_04=11, signalID=0)>
<NMEA(BDGSV, numMsg=2, msgNum=2, numSV=8, svid_01=33, elv_01=21.0, az_01=211, cno_01=27, svid_02=34, elv_02=18.0, az_02=153, cno_02=30, svid_03=41, elv_03=36.0, az_03=267, cno_03=22, svid_04=44, elv_04=35.0, az_04=96, cno_04=13, signalID=0)>
<NMEA(GLGSV, numMsg=2, msgNum=1, numSV=8, svid_01=66, elv_01=32.0, az_01=208, cno_01=31, svid_02=86, elv_02=27.0, az_02=97, cno_02=35, svid_03=73, elv_03=43.0, az_03=44, cno_03=, svid_04=76, elv_04=65.0, az_04=20, cno_04=, signalID=0)>

714 messages read.

Processing Complete
```

Unfortunately, getting this to work in PyGPSClient will also require very minor changes to the `pyubx2` and `pygpsclient` libraries to recognise the "BD" talker ID. With these changes made, PyGPSClient can also successfully parse BDGSV sentences:

<img width="1352" alt="Screenshot 2023-10-04 at 20 23 40" src="https://github.com/semuconsulting/pynmeagps/assets/28569967/a6a4f1bf-eda6-43d4-8795-d09efc7dacbf">


I'll endeavour to get these changes into the next releases of `pynmeagps`, `pyubx2` and `pygpsclient` in the next couple of weeks or so. See #43 .

**Reply by szilvasyz** [reply]
Hi @semuadmin,

On my locally stored Python files I have made the minor modifications you committed in the project files regarding talker ID "BD". Now it works like a charm also from online GPS connection, of course. I hope, the testing would not find any negative side-effect of this hack and we will see the realization in the future releases.

Thank you very much for quick response!

**Reply by semuadmin** [reply]
Hi @szilvasyz

The changes are now generally available in PyGPSClient v1.4.4.
